dart/dynamics/Entity.h

```diff
@@ -189,6 +198,19 @@ class Entity : public virtual common::Subject
   bool needsAccelerationUpdate() const;
 
 protected:
+
+  /// Used when constructing a Frame class, because the Frame constructor will
+  /// take care of setting up the parameters you pass into it
+  enum ConstructFrame_t { ConstructFrame };
+
+  explicit Entity(ConstructFrame_t);
```

Comment:
That's the jist of it. The full motivation for this is a bit more complicated, though:

In order to have decent inheritance structures, ``Entity`` and ``Frame`` need to be virtually inherited. This means that every derived class needs to call their constructors. This tends to be gross, confusing, and annoying, because many of those derived classes are pure abstract, which means they *need* to call some ``Frame`` & ``Entity`` constructor in the code even though the class itself can never be created and will therefore never *actually* call those constructors in practice.

One solution would be to have default (zero-argument) constructors for ``Frame`` & ``Entity``, but it's very important that every concrete derived class calls the correct constructor for ``Frame`` and ``Entity``, and having a default constructor for them would make concrete derived classes vulnerable to not realizing that they need to call a constructor, because by default it will just call their default constructors without any complaint.

So the solution I arrived at was to have some constructors that are easy and clear to call in derived classes. Using ``Entity(ConstructAbstract_t)`` is basically your way of saying "This class should never ever be instantiated, so calling the Entity constructor here is worthless since a more-derived class is going call it anyway." This will print out an error message if the constructor is ever actually called.

And then using ``Entity(ConstructFrame_t)`` is how you say "This class is inheriting from ``Frame``, and the ``Frame`` constructor will be used to set up the ``Entity`` properties."

Similar logic motivates the ``Frame(ConstructAbstract_t)`` constructor.

You can find this technique (overloading functions and constructors by using token enum types) in some Eigen classes. A specific example that comes to mind is the resize function:

```
enum NoChange_t { NoChange };

inline void resize(Index nbRows, Index nbCols);
inline void resize(Index nbRows, NoChange_t);
inline void resize(NoChange_t, Index nbCols);
```